test/storage/garbage_collector_test.cpp

```diff
@@ -671,4 +671,894 @@ TEST_F(GarbageCollectorTests, InsertUpdate1) {
     EXPECT_EQ(std::make_pair(2u, 0u), gc.PerformGarbageCollection());
   }
 }
+
+/*
+ * Consider performing GC on the version chain sorted from newest to oldest:
+ * DELETE
+ * U3
+ * U2
+ * U1
+ * INSERT
+ * T0 <- active txn
+ */
+// NOLINTNEXTLINE
+TEST_F(GarbageCollectorTests, SingleOLAP) {
+  for (uint32_t iteration = 0; iteration < num_iterations_; ++iteration) {
+    transaction::TransactionManager txn_manager(&buffer_pool_, true, LOGGING_DISABLED);
+    GarbageCollectorDataTableTestObject tested(&block_store_, max_columns_, &generator_);
+    storage::GarbageCollector gc(&txn_manager);
+
+    // T0
+    auto *txn0 = txn_manager.BeginTransaction();
+
+    // INSERT
+    auto *txn1 = txn_manager.BeginTransaction();
+    auto *insert_tuple = tested.GenerateRandomTuple(&generator_);
+    storage::TupleSlot slot = tested.table_.Insert(txn1, *insert_tuple);
+    txn_manager.Commit(txn1, TestCallbacks::EmptyCallback, nullptr);
+
+    tested.SelectIntoBuffer(txn0, slot);
+    EXPECT_FALSE(tested.select_result_);
+
+    // U1
+    auto *txn2 = txn_manager.BeginTransaction();
+    storage::ProjectedRow *update = tested.GenerateRandomUpdate(&generator_);
+    tested.table_.Update(txn2, slot, *update);
+    txn_manager.Commit(txn2, TestCallbacks::EmptyCallback, nullptr);
+
+    // U2
+    auto *txn3 = txn_manager.BeginTransaction();
+    update = tested.GenerateRandomUpdate(&generator_);
+    tested.table_.Update(txn3, slot, *update);
+    txn_manager.Commit(txn3, TestCallbacks::EmptyCallback, nullptr);
+
+    // U3
+    auto *txn4 = txn_manager.BeginTransaction();
+    update = tested.GenerateRandomUpdate(&generator_);
+    tested.table_.Update(txn4, slot, *update);
+    txn_manager.Commit(txn4, TestCallbacks::EmptyCallback, nullptr);
+
+    // U1, U2 will be unlinked. Can't unlink U3 as it is version chain head and T0 is still active
+    // Can't unlink INSERT as it is an Insert Undo Record
+    EXPECT_EQ(std::make_pair(0u, 2u), gc.PerformGarbageCollection());
+
+    tested.SelectIntoBuffer(txn0, slot);
+    EXPECT_FALSE(tested.select_result_);
+
+    txn_manager.Commit(txn0, TestCallbacks::EmptyCallback, nullptr);
+    // Unlink U4 as T0 committed and unlink read-only txn T0 and unlink INSERT as no active txn
+    // Deallocate U1, U2
+    EXPECT_EQ(std::make_pair(2u, 3u), gc.PerformGarbageCollection());
+    // Deallocate INSERT, U3
+    EXPECT_EQ(std::make_pair(2u, 0u), gc.PerformGarbageCollection());
+    // Nothing should be deallocated
+    EXPECT_EQ(std::make_pair(0u, 0u), gc.PerformGarbageCollection());
```

Comment:
Sorry, I am a little confused. There are five transactions and only four are deallocated. T0 is not deallocated according to your comment, why?

Reply:
Is it because T0 is read-only?

Reply:
Yeah read-only transactions don't get deallocated

Reply:
I saw you comment this in the later tests, but I think you should also comment it here.